crates/sui-indexer/src/bin/indexer_data_validation.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cmp::max;
+
+use anyhow::Result;
+use clap::Parser;
+use tracing::{error, info, warn};
+
+use sui_indexer::new_rpc_client;
+use sui_json_rpc_types::{
+    CheckpointId, ObjectChange, SuiObjectDataOptions, SuiTransactionResponseOptions,
+};
+use sui_sdk::SuiClient;
+use sui_types::base_types::{ObjectID, SequenceNumber};
+use sui_types::digests::TransactionDigest;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    info!("Running correctness check for indexer...");
+    let config = TestConfig::parse();
+    let fn_rpc_client = new_rpc_client(&config.fn_rpc_client_url).await?;
+    let indexer_rpc_client = new_rpc_client(&config.indexer_rpc_client_url).await?;
+
+    let fn_latest_checkpoint = fn_rpc_client
+        .read_api()
+        .get_latest_checkpoint_sequence_number()
+        .await?;
+    let indexer_latest_checkpoint = indexer_rpc_client
+        .read_api()
+        .get_latest_checkpoint_sequence_number()
+        .await?;
+
+    let checkpoint = config.start_checkpoint;
+    if checkpoint <= fn_latest_checkpoint && checkpoint <= indexer_latest_checkpoint {
+        // TODO(gegaowp): think of schedule and coordination with checkpoint and epoch,
+        // now try to check past 10 checkpoints.
+        for checkpoint in max(checkpoint - 10, 0)..checkpoint {
+            check_checkpoint(&config, &fn_rpc_client, &indexer_rpc_client, checkpoint).await?;
+        }
+    } else {
+        error!("Start checkpoint is not available in both FN and Indexer, start: {}, FN latest: {}, indexer latest: {}", 
+        checkpoint, fn_latest_checkpoint, indexer_latest_checkpoint);
+    }
+    Ok(())
+}
+
+pub async fn check_checkpoint(
+    config: &TestConfig,
+    fn_client: &SuiClient,
+    indexer_client: &SuiClient,
+    checkpoint: u64,
+) -> Result<(), anyhow::Error> {
+    let fn_checkpoint = fn_client
+        .read_api()
+        .get_checkpoint(CheckpointId::SequenceNumber(checkpoint))
+        .await?;
+    let indexer_checkpoint = indexer_client
+        .read_api()
+        .get_checkpoint(CheckpointId::SequenceNumber(checkpoint))
+        .await?;
+
+    if fn_checkpoint != indexer_checkpoint {
+        error!("Checkpoint mismatch found in {}", checkpoint);
+        warn!(
+            "Checkpoint mismatch with checkpoint seq {:?}:\nFN:\n{:?}\nIndexer:\n{:?} ",
+            checkpoint, fn_checkpoint, indexer_checkpoint
+        );
```

Comment:
should we just pick one?

Reply:
I wanted to use them differently:
- error can be used as "title" or counts only
- warn is for deeper investigations